JAVA-3626 and remaining pieces of JAVA-3696

jyemin opened the pull request:
Spec changes listed here: https://jira.mongodb.org/browse/DRIVERS-640

Work includes:

* Add initial ServerDescription to InternalConnection
* DefaultServerMonitor implements the new streaming isMaster protocol,
  if available, to detect topology changes sooner. This capability was
  added in MongoDB release 4.4.
* This requires the ability to increase the read timeout on a per-read
  basis.  Since the Stream class is public, this has to be done carefully
  in order to avoid using the streaming protocol with Stream implementations
  that don't have this ability.  All the built-in Stream implementations have
  had this ability added, so in practice it should not happen unless an
  application has created its own Stream implementation (unlikely).
* Implement the new server discovery and monitoring specification integration
  tests
* Fix some remaining bugs in error handling


Evergreen patch: https://evergreen.mongodb.com/version/5ebb4cf61e2d1701465a6a9c

There's something going wrong with async TLS that I will look into tomorrow.  That's why all the TLS tests are red.

jyemin commented on driver-core/src/main/com/mongodb/connection/Stream.java:
```diff
@@ -60,6 +60,42 @@
      */
     ByteBuf read(int numBytes) throws IOException;
 
+    /**
```
This was necessary because Stream is part of the public API, and on the off chance that someone is using a custom implementation of Stream, this will prevent us from using the streaming protocol until that implementation adds support for adding a per-read additional timeout.  All of the built-in implementations will return true for this.

jyemin commented on driver-core/src/main/com/mongodb/internal/async/client/AsyncMongoClient.java:
```diff
@@ -233,4 +236,19 @@
     <TResult> AsyncChangeStreamIterable<TResult> watch(AsyncClientSession clientSession, List<? extends Bson> pipeline,
                                                        Class<TResult> resultClass);
 
+    /**
```
This method is now used in the test runner, so it was need in SyncMongoClient, and therefore, here as well.

jyemin commented on driver-core/src/main/com/mongodb/internal/connection/AbstractMultiServerCluster.java:
```diff
@@ -281,7 +279,7 @@ private boolean handleReplicaSetMemberChanged(final ServerDescription newDescrip
                                 maxSetVersion, maxElectionId));
                     }
-                    addressToServerTupleMap.get(newDescription.getAddress()).server.invalidate();
+                    addressToServerTupleMap.get(newDescription.getAddress()).server.resetToConnecting();
```
According to the `updateRSFromPrimary` subroutine in SDAM, this path should only `replace description with a default ServerDescription of type "Unknown"`

ShaneHarvey replied:
Should `resetToConnecting` save the `newDescription`'s topologyVersion? Does it matter?

jyemin replied:
I don't think it matters, but if you think it does it should be in the spec.

ShaneHarvey replied:
Agreed. Please ignore.

jyemin commented on driver-core/src/main/com/mongodb/internal/connection/AbstractMultiServerCluster.java:
```diff
@@ -367,7 +365,7 @@ private void invalidateOldPrimaries(final ServerAddress newPrimary) {
                     LOGGER.info(format("Rediscovering type of existing primary %s", serverTuple.description.getAddress()));
                 }
-                serverTuple.server.invalidate(new MongoNotPrimaryException(new BsonDocument(), serverTuple.description.getAddress()));
+                serverTuple.server.invalidate();
```
According to SDAM:

> A note on invalidating the old primary: when a new primary is discovered, the client finds the previous primary (there should be none or one) and replaces its description with a default ServerDescription of type "Unknown." A multi-threaded client MUST request an immediate check for that server as soon as possible.

> If the old primary server version is 4.0 or earlier, the client MUST clear its connection pool for the old primary, too: the connections are all bad because the old primary has closed its sockets. If the old primary server version is 4.2 or newer, the client MUST NOT clear its connection pool for the old primary.

And that's what `invalidate` now does.  It doesn't make sense to forge an exception here, so got rid of that.  That lets us simplify the `ClustererableServer` API

ShaneHarvey replied:
Does invalidate request an immediate check like the spec says in your linked comment?

jyemin replied:
Yes

jyemin commented on driver-core/src/main/com/mongodb/internal/connection/CommandHelper.java:
```diff
@@ -19,8 +19,8 @@
 import com.mongodb.internal.async.SingleResultCallback;
-import com.mongodb.internal.validator.NoOpFieldNameValidator;
 import com.mongodb.internal.session.SessionContext;
+import com.mongodb.internal.validator.NoOpFieldNameValidator;
```
I'll revert this.

jyemin replied:
Done

jyemin commented on driver-core/src/main/com/mongodb/internal/connection/CommandMessage.java:
```diff
@@ -186,11 +203,14 @@ private int getOpMsgFlagBits() {
+        if (!requireOpMsgResponse()) {
+            flagBits = 1 << 1;
+        }
+        if (exhaustAllowed) {
```
This is the new bit: we now have to support the exhaustAllowed bit for streamable isMaster

jyemin commented on driver-core/src/main/com/mongodb/internal/connection/DefaultServer.java:
```diff
@@ -126,52 +127,46 @@ public ServerDescription getDescription() {
-
-    @Override
-    public void invalidate(final Throwable t, final int connectionGeneration, final int maxWireVersion) {
+    public void invalidate(final ConnectionState connectionState, final Throwable t, final int connectionGeneration,
```
By adding ConnectionState, we can share this method for both the before and after handshake behaviors, as there is only one small difference between them.

jyemin commented on driver-core/src/main/com/mongodb/internal/connection/DefaultServer.java:
```diff
@@ -126,52 +127,46 @@ public ServerDescription getDescription() {
             }
-            if ((t instanceof MongoSocketException && !(t instanceof MongoSocketReadTimeoutException))) {
-                invalidate();
+            if (t instanceof MongoSocketException
```
This now more closely follows the pseudocode in the spec

jyemin commented on driver-core/src/main/com/mongodb/internal/connection/DefaultServer.java:
```diff
@@ -200,15 +195,6 @@ private boolean isStale(final TopologyVersion currentTopologyVersion, final Topo
         return false;
     }
 
-    public void invalidate(final Throwable t, final int connectionGeneration, final int maxWireVersion,
```
Inlined this method, as it was a bit confusing to have another invalidate method

rozza replied:
Ack

jyemin commented on driver-core/src/main/com/mongodb/internal/connection/DefaultServerMonitor.java:
```diff
@@ -96,108 +113,137 @@ public void connect() {
-                        averageRoundTripTime.reset();
-                        currentServerDescription = getConnectingServerDescription(t);
-                    }
+            ServerDescription currentServerDescription = getConnectingServerDescription(null);
```
Compare this to the new pseudocode in the specification.  It now follows it much more closely, which simplified the logic a bit.

jyemin commented on driver-core/src/main/com/mongodb/internal/connection/DefaultServerMonitor.java:
```diff
@@ -96,108 +113,137 @@ public void connect() {
+                long start = System.nanoTime();
+                try {
+                    SessionContext sessionContext = new ClusterClockAdvancingSessionContext(NoOpSessionContext.INSTANCE, clusterClock);
+                    if (!connection.hasMoreToCome()) {
```
This part looks a bit different that the pseudocode in the spec, but might be a bit confusing.  The key point is that we're separating `send` and `receive` regardless of whether we're doing the streaming protocol.   The first time we get here, `moreToCome` will be false, so we send an isMaster, possibly using streaming protocol.  Then we immediately receive the first response.  The second time we get here (from the main loop above), `moreToCome` will be true if we're streaming, and we won't send, we'll just receive.  But if we aren't streaming, `moreToCome` will be false, and we'll both send and receive.

jyemin commented on driver-core/src/main/com/mongodb/internal/connection/DefaultServerMonitor.java:
```diff
@@ -96,108 +113,137 @@ public void connect() {
+                    MessageSettings.builder()
+                            .maxWireVersion(connection.getDescription().getMaxWireVersion())
+                            .build(),
+                    true);
```
`true` indicates `exhaustAllowed`.  According to OP_MSG protocol, it's fine to always set this to `true` even for old servers.  It will just be ignored if streaming isn't supported.

jyemin replied:
I tried it, and you're right, the server is not happy with the combination:

    {
      "ok": 0.0, 
     "errmsg": "An isMaster request with exhaust must specify 'maxAwaitTimeMS'", 
     "code": 51756, 
     "codeName": "Location51756"
    }

Do you think the server is acting according to specification here? If so, should I fix it or just assume that topologyVersion will always be available?

ShaneHarvey replied:
> Do you think the server is acting according to specification here?

I can see both sides to that debate. I suggest not sending exhaustAllowed if it's not a streaming isMaster command to avoid any unexpected behavior.

Would you like to open a server ticket or shall I? It might be useful to refer back to this in the future.

jyemin replied:
Done.  Stopped sending exhaustAllowed unless we are streaming.

I haven't opened a server ticket.  The current behavior seems acceptable.

jyemin commented on driver-core/src/main/com/mongodb/internal/connection/DefaultServerMonitor.java:
```diff
@@ -230,6 +276,13 @@ private long waitForSignalOrTimeout() throws InterruptedException {
+        public void cancelCurrentCheck() {
+            if (connection != null) {
+                currentCheckCancelled = true;
+                connection.close();
```
This is the only way to safely interrupt a read in Java: close the socket.

jyemin commented on driver-core/src/main/com/mongodb/internal/connection/DefaultServerMonitor.java:
```diff
@@ -272,27 +325,87 @@ static boolean shouldLogStageChange(final ServerDescription previous, final Serv
     }
+
+
+    private class RoundTripTimeRunnable implements Runnable {
```
This is the RTT monitor.

jyemin commented on driver-core/src/main/com/mongodb/internal/connection/ExponentiallyWeightedMovingAverage.java:
```diff
@@ -30,11 +27,11 @@
     }
 
-    void reset() {
+    synchronized void reset() {
```
Since both the RTT and Server Monitor share an instance of this class, we now have to make it threadsafe

rozza replied:
Ack

jyemin commented on driver-core/src/main/com/mongodb/internal/connection/InternalConnection.java:
```diff
@@ -34,6 +35,13 @@
+     *
+     * @return the initial server description
+     */
+    ServerDescription getInitialServerDescription();
```
This lets us support the requirement to update the cluster description based in the ismaster response from the initial handshake

jyemin commented on driver-core/src/main/com/mongodb/internal/connection/InternalConnectionInitializer.java:
```diff
@@ -17,12 +17,11 @@
 interface InternalConnectionInitializer {
 
-    ConnectionDescription initialize(InternalConnection internalConnection);
+    InternalConnectionInitializationDescription initialize(InternalConnection internalConnection);
```
Now we need both a ConnectionDescription and a ServerDescription, which are not quite the same thing

jyemin commented on driver-core/src/main/com/mongodb/internal/connection/InternalStreamConnection.java:
```diff
@@ -291,23 +346,32 @@ private void sendCommandMessage(final CommandMessage message,
-            return getCommandResult(decoder, responseBuffers, message.getId());
-        } finally {
-            responseBuffers.close();
+            hasMoreToCome = responseBuffers.getReplyHeader().hasMoreToCome();
```
This is the main part of the new support the the moreToCome bit

rozza replied:
Ack

jyemin commented on driver-core/src/test/unit/com/mongodb/internal/connection/AbstractServerDiscoveryAndMonitoringTest.java:
```diff
@@ -105,10 +107,10 @@ protected void applyApplicationError(final BsonDocument applicationError) {
         switch (when) {
             case "beforeHandshakeCompletes":
-                server.invalidate(errorGeneration);
+                server.invalidate(BEFORE_HANDSHAKE, exception, errorGeneration, maxWireVersion);
```
Now we use the same invalidate method for both cases, so updating this test to match

jyemin commented on driver-sync/src/test/functional/com/mongodb/client/AbstractRetryableReadsTest.java:
```diff
@@ -137,6 +138,12 @@ public void apply(final SocketSettings.Builder builder) {
                         builder.readTimeout(5, TimeUnit.SECONDS);
                     }
                 })
+                .applyToServerSettings(new Block<ServerSettings.Builder>() {
```
Shane suggested this to make these tests run faster.

rozza replied:
Ack

jyemin commented on driver-sync/src/test/functional/com/mongodb/client/TestServerListener.java:
```diff
@@ -0,0 +1,96 @@
+
+public class TestServerListener implements ServerListener {
+    private final List<ServerDescriptionChangedEvent> events = new ArrayList<>();
+    private final Lock lock = new ReentrantLock();
```
Stole this pattern from the CMAP test listener

jyemin commented on driver-sync/src/test/functional/com/mongodb/client/AbstractUnifiedTest.java:
```diff
@@ -386,9 +439,73 @@ public Void execute() {
+                        assertNull(failPoint);
                         failPoint = new TargetedFailPoint(operation);
                         failPoint.executeFailPoint();
+                    } else if (operationName.equals("configureFailPoint")) {
```
This supports all the new operations required by the new SDAM integration test runner.

ShaneHarvey commented on driver-core/src/main/com/mongodb/internal/connection/DefaultServerMonitor.java:
```diff
@@ -96,108 +113,137 @@ public void connect() {
+                if (connection == null || connection.isClosed()) {
+                    connection = internalConnectionFactory.create(serverId);
+                    connection.open();
+                    averageRoundTripTime.addSample(connection.getInitialServerDescription().getRoundTripTimeNanos());
```
Does the `averageRoundTripTime` still get reset on isMaster/network errors?

jyemin replied:
It doesn't.  Good catch.  Think we should add this sort of thing to the pseudocode?

ShaneHarvey replied:
Yes. Opened: https://jira.mongodb.org/browse/SPEC-1690

jyemin replied:
Done.  Still need to do the test though

rozza replied:
Is this test now done?

jyemin replied:
Yes, the prose test has been implemented.

jyemin commented on driver-core/src/main/com/mongodb/internal/connection/DefaultServerMonitor.java:
```diff
@@ -96,108 +113,137 @@ public void connect() {
+                    connection = internalConnectionFactory.create(serverId);
+                    connection.open();
+                    averageRoundTripTime.addSample(connection.getInitialServerDescription().getRoundTripTimeNanos());
+                    return connection.getInitialServerDescription();
```
I'm only inclined to change it if the spec is clear about it.  I don't think these events are particularly important to be honest.

ShaneHarvey replied:
My view is that the handshake fits the description of `ServerHeartbeatStartedEvent` (it is just a normal isMaster command after all). It looks like this was already the behavior in Java so it's not a regression. I'll defer to you.

jyemin replied:
I don't think it's worth the complication to change it.

jyemin commented on driver-core/src/main/com/mongodb/internal/connection/DefaultServerMonitor.java:
```diff
@@ -96,108 +113,137 @@ public void connect() {
+        private volatile InternalConnection connection = null;
+        private volatile boolean currentCheckCancelled;
+
+        void close() {
```
I don't think it matters.  Both end up closing the connection, and there's no reason to set the currectCheckCancelled boolean if the monitor is already closed.

ShaneHarvey replied:
SGTM.

jyemin commented:
Update: I've confirmed that the problem with the async TLS tests is in the tls-channel vendored code.  In all the other configurations (except Netty, still have to test that), closing the Stream causes pending read operation to immediately fail and throw an exception.  But with tls-channel, the pending read operation appears to hang indefinitely if the channel is closed while waiting for the read to complete.

This is a problem for DefaultServerMonitor.cancelCurrentCheck, which is required to force the pending isMaster call to be interrupted.  AFAIK there is no other way besides closing the socket/channel to get this to happen.  So I think we will need to investigate this and hopefully fix it in the tls-channel library.

CC @rozza

rozza commented:
Is there a new build on evergreen for this?

jyemin commented:
Just started one: https://evergreen.mongodb.com/version/5ec541db3066151c2c8285ce

I'm expecting it to fail with some timeouts.

jyemin commented:
New commits, and a new, greener, patch: https://evergreen.mongodb.com/version/5ec5a65f9ccd4e743f7f1d53

jyemin commented on driver-core/src/main/com/mongodb/internal/connection/DefaultServerMonitor.java:
```diff
@@ -159,7 +159,7 @@ public void run() {
                 }
 
-                if (((connection == null || connection.supportsAdditionalTimeout())
+                if (((connection == null || shouldStreamResponses(currentServerDescription))
```
This fixes the timeouts on Evergreen, which were caused by this bug interacting with a test bug:

* the bug here is that we have to use the same condition here as we do to decide whether streaming will be done.  Without this change, we won't stream but we also won't sleep.
* there is a bug in reactive client side encryption tests where we don't close the MongoClient after each test.  This causes us to have hundreds of open MongoClient instances 

The combination of the two means that with TLS (which won't stream, due to the TSL library issue) we will have hundreds of MongoClient instances, all with monitors polling in a tight loop.

jyemin commented:
I took this out of draft now that Evergreen looks good.  I still have the two prose tests to implement, and I will get to those ASAP.

jyemin commented:
Added the prose tests from https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/tests/README.rst

The patch build is a bit red because I neglected to skip the new RTT prose test if server version is older than 4.4.  I added that check in the commit though before I pushed it, but didn't bother to re-run the Evergreen patch.

Evergreen: https://evergreen.mongodb.com/version/5ec67912e3c331477ce4b84e

jyemin commented:
@ShaneHarvey I think I've resolved all of your comment threads.  PTAL

rozza commented on driver-core/src/main/com/mongodb/internal/connection/DefaultServerMonitor.java:
```diff
@@ -75,12 +87,17 @@
         monitorThread = new Thread(monitor, "cluster-" + this.serverId.getClusterId() + "-" + this.serverId.getAddress());
         monitorThread.setDaemon(true);
+        roundTripTimeMonitor = new RoundTripTimeRunnable();
+        roundTripTimeMonitorThread = new Thread(roundTripTimeMonitor,
```
Maybe something for the future - but having an easy way to remove the rtt logs from the main logs when debugging would be good / something to document.

jyemin replied:
`RoundTripTimeRunnable` doesn't currently log.  It does get a unique thread name though.

rozza replied:
ACK

rozza commented on driver-core/src/main/com/mongodb/internal/connection/InternalStreamConnection.java:
```diff
@@ -112,20 +118,33 @@ public InternalStreamConnection(final ServerId serverId, final StreamFactory str
         try {
             stream.open();
-            description = connectionInitializer.initialize(this);
+            LOGGER.debug("Done opening stream to " + serverId.toString());
```
Is this a debugging log statement?  

nit: `if (LOGGER.isDebugEnabled()) {`

jyemin replied:
Removed

rozza commented on driver-core/src/main/com/mongodb/internal/connection/InternalStreamConnection.java:
```diff
@@ -112,20 +118,33 @@ public InternalStreamConnection(final ServerId serverId, final StreamFactory str
+            initialServerDescription = initializationDescription.getServerDescription();
             opened.set(true);
             sendCompressor = findSendCompressor(description);
             LOGGER.info(format("Opened connection [%s] to %s", getId(), serverId.getAddress()));
```
nit: Add isInfoEnabled check.

jyemin replied:
Done

rozza commented on driver-core/src/test/unit/com/mongodb/internal/connection/TestConnectionPoolListener.java:
```diff
@@ -89,7 +89,7 @@ public TestConnectionPoolListener() {
 
     private <T> boolean containsEvent(final Class<T> eventClass, final int expectedEventCount) {
-        return countEvents(eventClass) == expectedEventCount;
+        return countEvents(eventClass) >= expectedEventCount;      // TODO
```
Whats the todo here?

jyemin replied:
I think it's correct as is with the change, as you might have more events than expected, just not less.  Removed  the TODO

jyemin commented:
@rozza ready for re-review

jyemin commented on driver-core/src/main/com/mongodb/internal/connection/DefaultServerMonitor.java:
```diff
@@ -229,13 +230,18 @@ private ServerDescription lookupServerDescription(final ServerDescription curren
                             .maxWireVersion(connection.getDescription().getMaxWireVersion())
                             .build(),
-                    true);
+                    shouldStreamResponses(serverDescription));
```
Went for the latter and changed the variable name

jyemin replied:
Done

jyemin commented on driver-core/src/main/com/mongodb/internal/connection/DefaultServerMonitor.java:
```diff
@@ -272,27 +344,88 @@ static boolean shouldLogStageChange(final ServerDescription previous, final Serv
+                                pingServer(connection);
+                            } catch (MongoSocketException | MongoCommandException e) {
+                                connection.close();
+                                connection = null;
```
Yes, this is a bug.  But instead of throwing e I instead removed this entire catch block and close the connection in the outer catch block.  This way if `connection.open` throws an exception, the connection will be set to null as well and re-created the next time around the loop.

jyemin commented on driver-core/src/main/com/mongodb/internal/connection/DefaultServerMonitor.java:
```diff
@@ -96,108 +112,151 @@ public void connect() {
-                        new ServerHeartbeatFailedEvent(connection.getDescription().getConnectionId(), System.nanoTime() - start, e));
-                throw e;
+                if (connection == null || connection.isClosed()) {
+                    currentCheckCancelled = false;
```
After talking with Divjot, decided that this variable should be reset here in order to avoid unnecessarily going into the wait state twice in a row after a cancel.

rozza replied:
Good catch

jyemin commented:
Evergreen patch for latest commit: https://evergreen.mongodb.com/version/5ecefe423066155144e6821f

jyemin commented:
Thanks everyone for the thorough reviews.  I'm going to close this, rebased on master, squash some commits, and merge it locally.
